AustinKellar/PowerOn-Prototype
# Cursor changes based on what the selected action and what the user is hovering

Goal: Give clear indications about what mouse clicks do by changing the texture of the cursor in certain situations

What I changed:
- Added 3 new assets in a new folder called Cursors:
   - default.png
   - hand.png
   - delete.png
   - invalid.png
- I picked 3 cursors from this site: https://icons8.com/icon/new-icons/all, from the same set (Windows 10) so that they all match in style
- New Script:
   - CursorController.cs
- The cursor texture changes dynamically based on the situation.
   - When the forest player is in 'Grow' mode and hovering either a branch or a tree, the Hand cursor is displayed
   - When the forest player is in 'Destroy' mode and hovering a branch, the Delete cursor is displayed
   - If the player is hovering over an invalid action (for example hovering a tree trunk in Destroy mode), the Invalid cursor is displayed
   - In all other situations, the Default cursor is displayed
- Since all of the cursors communicate a different spot where the click will happen, I changed the 'hotspot' of each one so that the clicking feels as intuitive as possible. I.E. the hotspot or click-point for the Hand cursor is at the tip of the finger, while the hotspot for the Delete cursor is at the center of the X.
- To improve quality of life, I also widened the clicking hitbox for branches. This was done by adding a second box-collider to the prefab which is set to Trigger mode. It can easily be adjusted, so let me know if you want the clicking hitbox any smaller or larger

How to Test:
- Ensure that you see the Default cursor ![icons8-cursor-50](https://user-images.githubusercontent.com/28333213/48589094-c59c1080-e8ee-11e8-9856-14a21635edb1.png) iff:
   - You are hovering over the anything besides tree trunks or branches (the background, squirrel or UI)

- Ensure that you see the Hand cursor ![icons8-hand-cursor-50](https://user-images.githubusercontent.com/28333213/48589132-03009e00-e8ef-11e8-851b-a59881672551.png) iff:
  - You are in 'Grow' mode and hovering over a tree trunk with sap available
  - You are in 'Grow' mode and hovering over a branch that is not yet max length and sap is available

- Ensure that you see the Destroy cursor ![icons8-delete-32](https://user-images.githubusercontent.com/28333213/48589045-8968b000-e8ee-11e8-9dda-c7b76b411c56.png) iff:
   - You are in 'Destroy' mode and hovering over a branch that is above the squirrel

- Ensure that you see the Invalid cursor ![icons8-cancel-48](https://user-images.githubusercontent.com/28333213/48597372-4587a200-e912-11e8-9393-fc746b649076.png) iff:
   - You are in 'Destroy' mode, hovering a tree trunk
   - You are in 'Destroy' mode, hovering a branch that is below the squirrel
   - You are in 'Grow' mode, hovering over a tree trunk with no sap
   - You are in 'Grow' mode, hovering over a branch that is already at its maximum length
- Make sure that branches are easier to click. If the hitbox is still not wide enough, let me know.
- Make sure that the cursor feels intuitive. Let me know if you think it should behave differently in any circumstance.

Notes:
- I added the Invalid cursor functionality after I first put the PR up and it made the code a lot longer and a bit messier. Refactor suggestions are welcome.
- It was pretty difficult for me to find icons that were easy to work with, free and matching in style so if you want to use different icons than the ones I found, I ask that you find some yourself and send them to me.
- Early on in my own testing, I noticed that the cursors would sometimes flicker when I move the mouse quickly. I think I fixed this problem by only setting the cursor texture if it changes, rather than every frame. Once I added in that change, the flickering stopped on my screen. Let me know if you see any flickering.
